# Configuring R and RStudio {.imageChapter}

<div class="chapter_image"><img src="media/ch_farm.jpg"></div>

## Packages

The main library for marker-based work presented in this text is the *gstudio* package.  This package was written specifically to allow the easy integration of genetic data into the standard R workflow such that your genotypes are simply another class of data within common R data containers such as `data.frame` and `list` objects.  

There is an older version of this library on CRAN, though the development has switched over to GitHub and you should the library from there.  To install the package, you can use the devtools library to directly as:

```{r eval=FALSE}
library(devtools)
install_github("dyerlab/gstudio")
```

In addition to this library, the text in this book utilizes requires several other packages from both CRAN and GitHub. The following if you run the following code, you should be up-to-date on the necessary packages used throughout this text.

```{r echo=FALSE, warning=FALSE}
files <- list.files(".",pattern=".Rmd")
libraries <- NA
for( file in files ) {
  suppressWarnings(s <- system( paste("grep -E -o 'library\\(\\w+\\)'",file), intern=TRUE))
  if( length(s) > 0 ) {
    s <- strsplit(s,"(",fixed=TRUE)
    for( item in s ){
      if( length(item) == 2){
        library <- strsplit(item[2],")",fixed=TRUE)[[1]]
        libraries <- c(libraries,library)
      }
    }
  }
}
pkgs <- sort( unique(libraries) )
pkgs <- pkgs[!is.na(pkgs)]
pkgs_df <- data.frame(Name = pkgs, Title = NA)
for(i in seq_along(pkgs)){
  f = system.file(package = pkgs[i], "DESCRIPTION")
  if( nchar(f)> 1) {
    # Title is always on 3rd line
    title = readLines(f)
    title = title[grep("Title: ", title)]
    pkgs_df$Title[i] = gsub("Title: ", "", title)    
  }
}
knitr::kable(pkgs_df,caption = "R packages used in the examples shown in this book.")

# make sure these libraries are already installed on this machine
inst_pkgs <- installed.packages()
to_install <- setdiff( pkgs, inst_pkgs )
if( length(to_install))
  install.packages(to_install, repos="https://cran.rstudio.org")


# See if maxent.jar is installed in the dismo java package.
path <- system.file("java", package="dismo")
files <- list.files(system.file("java", package="dismo"))
if( !("maxent.jar" %in% files) )
  system( paste("cp ./spatial_data/maxent.jar", path) )

```

### RGDAL & RGEOS {#rgdal-rgeos-packages}

Every time I upgrade in any significant way, two R libraries seem to raise their ugly heads and scream like a spoiled childâ€”rgdal and rgeos.  Why do these two have to be SOOOO much of a pain? Why can't we have a auto build of a binary with all the options in it for OSX? Who knows? I always feel like I get the fuzzy end of the lollipop with these two. Here is my latest approach for getting them going.

First you have to make sure you have the latest GDAL libraries. I used to get mine from [Kyngchaos](http://www.kyngchaos.com/software/frameworks), just download the framework, install it, and then do some kind of long R CMD INSTALL dance, which seems to no longer work for me. I also tried installing from Ripley's repository and found that (a) It was a version older than the one I already had on my machine, and (b) you can't install from that repository, there is a malformed header and the install.packages() function just barfs.

Time to try something new. I typically stay away from the various installer frameworks out there on OSX to keep everything in Frameworks. But this time, I used MacPorts. You can find the latest version here. Here is how I got it to help me out.  

- Download XCode from Apple, it is both free and has a lot of tools that make your life as a programmer easier.  It is a big package and you'll have to install the command line developer tools as well.  You will be prompted on how to do this.
- Downloaded the version of macports for your OS, I'm currently on 10.11 and installed it with little problems.  It takes a bit of time at the end of the installation because it is downloading a lot of information.  Be patient. 
- In the terminal, I updated it `sudo ports -v selfupdate` and again be patient, it is going to grab a lot of stuff from the internet.
- I then used it to install `gdal` as a unix library (rather than as a framework so it won't be located in /Library/Frameworks) by sudo ports install `gdal`. There were a lot of dependencies for this one so it took a while.
- I then had R install rgdal as `install.packages( rgdal, type="source")`

Worked like a charm.


## Data Sets

This course will use two different data sets from the Dyer Laboratory.

### The Sonoran desert bark beetle, *Araptus attenuata*

```{r echo=FALSE}
knitr::include_graphics("media/araptus.png")
```

*Arapatus attenuatus* is a bark beetle endemic to the Sonoran desert.  It is known only from the senescing stems of the desert plant, *Euphorbia lomelii* (syn *Pedilanthus macrocarpus*).  Samples for this data set were collected from 39 populations throughout the species range.

```{r message=FALSE, echo=FALSE, warning=FALSE, fig.cap="Sampling locales for the `arapat` data set."}
library(ggplot2)
library(ggmap)
library(gstudio)
library(ggrepel)
data(arapat)
coords <- strata_coordinates(arapat)
map <- population_map(coords)
ggmap(map) + geom_point(aes(x=Longitude,y=Latitude), data=coords) + xlab("Longitude") + ylab("Latitude") + geom_text_repel(aes(x=Longitude, y=Latitude, label=Stratum), data=coords) 
```



#### Nuclear Markers

The markers in the `arapat` data set consist of eight polymorphic and codominant nuclear markers. 

```{r message=FALSE}
library(gstudio)
data(arapat)
column_class( arapat, "locus") 
```

These markers span a range of allelic diversity and richness.

```{r echo=FALSE}
df <- merge( genetic_diversity(arapat,mode="A"), genetic_diversity(arapat,mode="He"))
df <- merge( df, genetic_diversity(arapat, mode="Ae"))
knitr::kable(df,align=c("l","c","c","c"),digits=3)
```



#### Methylation Markers

In addition to codominant nuclear markers, an msAFLP analysis was performed on the major clade and paired sequence and $C_pG$ methylation markers were derived.



### The Flowering Dogwood *Cornus florida*

```{r echo=FALSE, fig.cap="Inflorescence for flowering dogwood with conspicuous showy bracts and many small flowers in the center."}
knitr::include_graphics("media/dogwood_flower.png",)
```


Flowering dogwood is an ubiquitous feature of eastern north american forests. 

The data set contains five microsatellite loci.

```{r}
data(cornus)
column_class( cornus, "locus") 
```

of roughly equal genetic diversity.

```{r echo=FALSE}
df <- merge( genetic_diversity(cornus,mode="A"), genetic_diversity(cornus,mode="He"))
df <- merge( df, genetic_diversity(cornus, mode="Ae"))
knitr::kable(df,align=c("l","c","c","c"),digits=3)
```



