# Landscape Genetics {.imageChapter}

<div class="chapter_image"><img src="media/ch_farm.jpg"></div>

Landscape genetics may---or may not (see Dyer 2015)---be a coherient field of study.  The the term itself was coined by Manel *et al.* (2003).  In that paper, they define the breadth of what they consider landscape genetics to be as

> Landscape genetics can resolve population substructure across different geographical scales at fine taxonomic levels; thus, it is different from the existing disciplines of biogeography, which focuses mainly on species diversity patterns at broad temporal and spatial scales, and phylogeography, which combines phylogenetics and biogeography... 

and outline the specific approaches that are used in its execution

> The two key steps of landscape genetics are the detection of genetic discontinuities and the correlation of these discontinuities with landscape and environmental features, such as barriers (e.g. mountains, gradient of humidity).

In the intervening period, many research groups have grabbed onto the moniker and applied "landscape genetic" approaches to questions focusing on connectivity.  At a fundamental level, the entirely of landscape genetics is dependent upon a single characteristic equation

\[
G \propto f(E)
\]

where $G$ is a measure of spatial genetic structure, which was created and has been maintained by some function of one or more $E$ external (spatial, ecological, topological) features.  In fact, the distribution of genetic variance, $G$, can be considered the culmination of a stratigraphy of demographic and evolutionary processes including (ordered from oldest to most recent):  

* speciation
* phylogeography
* range expansion & isolation
* inbreeding
* contemporary genenetic connectivity

As such, a fundamental question in population genetics (*sensu lauto*) and the ultimate question in landscape genetics (*sensu stricto*) is to determine both the identity and relative magnitude of the external predictor variables in the equation above.  The spatial distribution of genotypes *is* the answer, we are left with the task of determining a plausable set of predictors---a task that this entire text has been designed to uncover. 

## R Toolkit

The R ecosystem has had far reaching influences on applied population genetic and landscape ecological endeavours.  It is being positioned as a fundamental tool in data science due in part to its open nature and community contributions.  In this text, R will be used specifically as a toolkit 

The main library for marker-based work presented in this text is the *gstudio* package.  This package was written specifically to allow the easy integration of genetic data into the standard R workflow such that your genotypes are simply another class of data within common R data containers such as `data.frame` and `list` objects.  

To install the package, you will need the *devtools* library.  Once installed, you can grab the most recent version of *gstudio* from github by:

```{r eval=FALSE}
library(devtools)
install_github("dyerlab/gstudio")
```

In addition to this library, the text in this book utilizes requires several other packages from both CRAN and GitHub. The following if you run the following code, you should be up-to-date on the necessary packages used throughout this text.

```{r echo=FALSE, warning=FALSE}
files <- list.files(path=c(".","./presentations/"),pattern=".Rmd")
libraries <- NA
for( file in files ) {
  suppressWarnings(s <- system( paste("grep -E -o 'library\\(\\w+\\)'",file), intern=TRUE))
  if( length(s) > 0 ) {
    s <- strsplit(s,"(",fixed=TRUE)
    for( item in s ){
      if( length(item) == 2){
        library <- strsplit(item[2],")",fixed=TRUE)[[1]]
        libraries <- c(libraries,library)
      }
    }
  }
}
pkgs <- sort( unique(libraries) )
pkgs <- pkgs[!is.na(pkgs)]
pkgs_df <- data.frame(Name = pkgs, Title = NA)
for(i in seq_along(pkgs)){
  f = system.file(package = pkgs[i], "DESCRIPTION")
  if( nchar(f)> 1) {
    # Title is always on 3rd line
    title = readLines(f)
    title = title[grep("Title: ", title)]
    pkgs_df$Title[i] = gsub("Title: ", "", title)    
  }
}
knitr::kable(pkgs_df,caption = "R packages used in the examples shown in this book.")

# make sure these libraries are already installed on this machine
inst_pkgs <- installed.packages()
to_install <- setdiff( pkgs, inst_pkgs )
if( length(to_install))
  install.packages(to_install, repos="https://cran.rstudio.org")


# See if maxent.jar is installed in the dismo java package.
path <- system.file("java", package="dismo")
files <- list.files(system.file("java", package="dismo"))
if( !("maxent.jar" %in% files) )
  system( paste("cp ./spatial_data/maxent.jar", path) )

```

Make sure these are installed and up-to-date.


## Data Sets

This course will use two different data sets from the Dyer Laboratory.

### The Sonoran desert bark beetle, *Araptus attenuata*

```{r echo=FALSE}
knitr::include_graphics("media/araptus.png")
```

*Arapatus attenuatus* is a bark beetle endemic to the Sonoran desert.  It is known only from the senescing stems of the desert plant, *Euphorbia lomelii* (syn *Pedilanthus macrocarpus*).  Samples for this data set were collected from 39 populations throughout the species range.

```{r message=FALSE, echo=FALSE, warning=FALSE, fig.cap="Sampling locales for the `arapat` data set."}
library(ggplot2)
library(ggmap)
library(gstudio)
library(ggrepel)
data(arapat)
coords <- strata_coordinates(arapat)
map <- population_map(coords)
ggmap(map) + geom_point(aes(x=Longitude,y=Latitude), data=coords) + xlab("Longitude") + ylab("Latitude") + geom_text_repel(aes(x=Longitude, y=Latitude, label=Stratum), data=coords) 
```



#### Nuclear Markers

The markers in the `arapat` data set consist of eight polymorphic and codominant nuclear markers. 

```{r message=FALSE}
library(gstudio)
data(arapat)
column_class( arapat, "locus") 
```

These markers span a range of allelic diversity and richness.

```{r echo=FALSE}
df <- merge( genetic_diversity(arapat,mode="A"), genetic_diversity(arapat,mode="He"))
df <- merge( df, genetic_diversity(arapat, mode="Ae"))
knitr::kable(df,align=c("l","c","c","c"),digits=3)
```



#### AFLP Markers

In addition to codominant nuclear markers, a set of AFLP markers have been used to perform gene scans in an attempt to quantify adaptive genetic variance.



### The Flowering Dogwood *Cornus florida*

```{r echo=FALSE, fig.cap="Inflorescence for flowering dogwood with conspicuous showy bracts and many small flowers in the center."}
knitr::include_graphics("media/dogwood_flower.png",)
```


Flowering dogwood is an ubiquitous feature of eastern north american forests. 

The data set contains five microsatellite loci.

```{r}
data(cornus)
column_class( cornus, "locus") 
```

of roughly equal genetic diversity.

```{r echo=FALSE}
df <- merge( genetic_diversity(cornus,mode="A"), genetic_diversity(cornus,mode="He"))
df <- merge( df, genetic_diversity(cornus, mode="Ae"))
knitr::kable(df,align=c("l","c","c","c"),digits=3)
```




