<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Landscape Genetic Data Analysis</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="A textbook for the use of R in spatial genetic data analysis.">
  <meta name="generator" content="bookdown <!--bookdown:version--> and GitBook 2.6.7">

  <meta property="og:title" content="Landscape Genetic Data Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A textbook for the use of R in spatial genetic data analysis." />
  <meta name="github-repo" content="dyerlab/AMG" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Landscape Genetic Data Analysis" />
  
  <meta name="twitter:description" content="A textbook for the use of R in spatial genetic data analysis." />
  

<meta name="author" content="Rodney J. Dyer">


<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="media/style.css" type="text/css" />
</head>

<body>


<!--bookdown:title:start-->
<div id="header">
<h1 class="title">Landscape Genetic Data Analysis</h1>
<h4 class="author"><em>Rodney J. Dyer</em></h4>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#frontmatter">Frontmatter</a></li>
<li><a href="#part-background">(PART) Background</a></li>
<li><a href="#configuring-r-and-rstudio"><span class="toc-section-number">1</span> Configuring R and RStudio</a><ul>
<li><a href="#packages"><span class="toc-section-number">1.1</span> Packages</a><ul>
<li><a href="#rgdal-rgeos-packages"><span class="toc-section-number">1.1.1</span> RGDAL &amp; RGEOS</a></li>
</ul></li>
<li><a href="#data-sets"><span class="toc-section-number">1.2</span> Data Sets</a><ul>
<li><a href="#the-sonoran-desert-bark-beetle-araptus-attenuata"><span class="toc-section-number">1.2.1</span> The Sonoran desert bark beetle, <em>Araptus attenuata</em></a></li>
<li><a href="#the-flowering-dogwood-cornus-florida"><span class="toc-section-number">1.2.2</span> The Flowering Dogwood <em>Cornus florida</em></a></li>
</ul></li>
</ul></li>
<li><a href="#vector-raster-data"><span class="toc-section-number">2</span> Vector &amp; Raster Data</a><ul>
<li><a href="#points"><span class="toc-section-number">2.1</span> Points</a></li>
<li><a href="#lines"><span class="toc-section-number">2.2</span> Lines</a></li>
<li><a href="#polygons"><span class="toc-section-number">2.3</span> Polygons</a></li>
<li><a href="#rasters"><span class="toc-section-number">2.4</span> Rasters</a></li>
<li><a href="#cropping-rasters"><span class="toc-section-number">2.5</span> Cropping Rasters</a><ul>
<li><a href="#cropping-rasters-via-polygons"><span class="toc-section-number">2.5.1</span> Cropping Rasters Via Polygons</a></li>
<li><a href="#cropping-rasters-via-convex-hull"><span class="toc-section-number">2.5.2</span> Cropping Rasters Via Convex Hull</a></li>
</ul></li>
<li><a href="#modifying-rasters"><span class="toc-section-number">2.6</span> Modifying Rasters</a></li>
<li><a href="#extracting-point-data-from-rasters"><span class="toc-section-number">2.7</span> Extracting Point Data From Rasters</a><ul>
<li><a href="#stacks-of-rasters"><span class="toc-section-number">2.7.1</span> Stacks of Rasters</a></li>
</ul></li>
<li><a href="#rasters-ggplot"><span class="toc-section-number">2.8</span> Rasters &amp; ggplot</a></li>
<li><a href="#d-visualization"><span class="toc-section-number">2.9</span> 3D Visualization</a></li>
<li><a href="#saving-exporting-vector-raster-objects"><span class="toc-section-number">2.10</span> Saving &amp; Exporting Vector &amp; Raster Objects</a></li>
</ul></li>
<li><a href="#map-projections"><span class="toc-section-number">3</span> Map Projections</a><ul>
<li><a href="#projections"><span class="toc-section-number">3.1</span> Projections</a><ul>
<li><a href="#equatorial-projections"><span class="toc-section-number">3.1.1</span> Equatorial Projections</a></li>
<li><a href="#azimuth-projections"><span class="toc-section-number">3.1.2</span> Azimuth Projections</a></li>
<li><a href="#polar-conic-projections"><span class="toc-section-number">3.1.3</span> Polar Conic Projections</a></li>
<li><a href="#miscellaneous-projections"><span class="toc-section-number">3.1.4</span> Miscellaneous Projections</a></li>
<li><a href="#reprojecting-rasters"><span class="toc-section-number">3.1.5</span> Reprojecting Rasters</a></li>
<li><a href="#projecting-in-ggplot"><span class="toc-section-number">3.1.6</span> Projecting In GGPlot</a></li>
</ul></li>
<li><a href="#coordinate-systems"><span class="toc-section-number">3.2</span> Coordinate Systems</a><ul>
<li><a href="#changing-datum"><span class="toc-section-number">3.2.1</span> Changing Datum</a></li>
</ul></li>
</ul></li>
<li><a href="#part-genetic-data">(PART) Genetic Data</a></li>
<li><a href="#genetic-markers-basic-summary-statistics"><span class="toc-section-number">4</span> Genetic Markers &amp; Basic Summary Statistics</a><ul>
<li><a href="#marker-types"><span class="toc-section-number">4.1</span> Marker Types</a></li>
<li><a href="#sampling"><span class="toc-section-number">4.2</span> Sampling</a></li>
<li><a href="#allele-frequencies"><span class="toc-section-number">4.3</span> Allele Frequencies</a></li>
<li><a href="#hardy-weinberg-equilibrium"><span class="toc-section-number">4.4</span> Hardy-Weinberg Equilibrium</a></li>
</ul></li>
<li><a href="#genetic-distance-diversity"><span class="toc-section-number">5</span> Genetic Distance &amp; Diversity</a></li>
<li><a href="#genetic-structure"><span class="toc-section-number">6</span> Genetic Structure</a></li>
<li><a href="#part-barriers-resistance">(PART) Barriers &amp; Resistance</a></li>
<li><a href="#remote-sensing-lidar"><span class="toc-section-number">7</span> Remote Sensing &amp; Lidar</a></li>
<li><a href="#barrier-detection"><span class="toc-section-number">8</span> Barrier Detection</a></li>
<li><a href="#resistance-approaches"><span class="toc-section-number">9</span> Resistance Approaches</a></li>
<li><a href="#part-spatial-structure">(PART) Spatial Structure</a></li>
<li><a href="#networks"><span class="toc-section-number">10</span> Networks</a></li>
<li><a href="#autocorrelation"><span class="toc-section-number">11</span> Autocorrelation</a></li>
<li><a href="#part-adaptive-variance">(PART) Adaptive Variance</a></li>
<li><a href="#adaptive-variance"><span class="toc-section-number">12</span> Adaptive Variance</a></li>
<li><a href="#quantitative-genetics"><span class="toc-section-number">13</span> Quantitative Genetics</a></li>
<li><a href="#genetic-distance-diversity-1"><span class="toc-section-number">14</span> Genetic Distance &amp; Diversity</a></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Landscape Genetic Data Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<div id="frontmatter" class="section level1 unnumbered white_letters">
<h1>Frontmatter</h1>
<p>Build Date: <strong>Wed Sep 14 13:42:39 2016</strong></p>
<p>The content of this text was created to assist in the teaching of a course entitled <em>Landscape Genetic Data Analysis</em>, organized by <a href="http://prstatistics.com">PRStatistics</a>. All content for this book is contained on Github at <a href="https://github.com/dyerlab/applied_landscape_genetics" class="uri">https://github.com/dyerlab/applied_landscape_genetics</a>.</p>
<p><a href="http://dyerlab.bio.vcu.edu"> <img src="media/LGDA-Titlepage-Small.png" width="100" /> </a></p>
<p>© 2016 by R.J. Dyer.</p>
<p><a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>. Under this license, authorized individuals may copy, distribute, display and perform the work and make derivative works and remixes based on this text only if they give the original author credit (attribution). You are also free to distribute derivative works only under a license identical (“not more restrictive”) to the license that governs this original work.</p>
<p>Dr. Rodney Dyer is an Associate Professor in the Department of Biology and the Assistant Director for the Center for Environmental Studies at Virginia Commonwealth University in Richmond, Virginia, USA. His research focuses on genetic connectivity and structure and how the environment influences both. More information on his research can be found at <a href="http://dyerlab.bio.vcu.edu" class="uri">http://dyerlab.bio.vcu.edu</a>.</p>
<!--chapter:end:index.Rmd-->
</div>
<div id="part-background" class="section level1 unnumbered">
<h1>(PART) Background</h1>
<!--chapter:end:part_day_1.Rmd-->
</div>
<div id="configuring-r-and-rstudio" class="section level1 imageChapter">
<h1><span class="header-section-number">1</span> Configuring R and RStudio</h1>
<div class="chapter_image">
<img src="media/ch_farm.jpg">
</div>
<div id="packages" class="section level2">
<h2><span class="header-section-number">1.1</span> Packages</h2>
<p>The main library for marker-based work presented in this text is the <em>gstudio</em> package. This package was written specifically to allow the easy integration of genetic data into the standard R workflow such that your genotypes are simply another class of data within common R data containers such as <code>data.frame</code> and <code>list</code> objects.</p>
<p>There is an older version of this library on CRAN, though the development has switched over to GitHub and you should the library from there. To install the package, you can use the devtools library to directly as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;dyerlab/gstudio&quot;</span>)</code></pre></div>
<p>In addition to this library, the text in this book utilizes requires several other packages from both CRAN and GitHub. The following if you run the following code, you should be up-to-date on the necessary packages used throughout this text.</p>
<table>
<caption>(#tab:unnamed-chunk-3)R packages used in the examples shown in this book.</caption>
<thead>
<tr class="header">
<th align="left">Name</th>
<th align="left">Title</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">cowplot</td>
<td align="left">Streamlined Plot Theme and Plot Annotations for ‘ggplot2’</td>
</tr>
<tr class="even">
<td align="left">devtools</td>
<td align="left">Tools to Make Developing R Packages Easier</td>
</tr>
<tr class="odd">
<td align="left">ggmap</td>
<td align="left">Spatial Visualization with ggplot2</td>
</tr>
<tr class="even">
<td align="left">ggplot2</td>
<td align="left">An Implementation of the Grammar of Graphics</td>
</tr>
<tr class="odd">
<td align="left">ggrepel</td>
<td align="left">Repulsive Text and Label Geoms for ‘ggplot2’</td>
</tr>
<tr class="even">
<td align="left">gstudio</td>
<td align="left">Tools Related to the Spatial Analysis of Genetic Marker Data</td>
</tr>
<tr class="odd">
<td align="left">maps</td>
<td align="left">Draw Geographical Maps</td>
</tr>
<tr class="even">
<td align="left">raster</td>
<td align="left">Geographic Data Analysis and Modeling</td>
</tr>
<tr class="odd">
<td align="left">rasterVis</td>
<td align="left">Visualization Methods for Raster Data</td>
</tr>
<tr class="even">
<td align="left">rgdal</td>
<td align="left">Bindings for the Geospatial Data Abstraction Library</td>
</tr>
<tr class="odd">
<td align="left">rgeos</td>
<td align="left">Interface to Geometry Engine - Open Source (GEOS)</td>
</tr>
<tr class="even">
<td align="left">sp</td>
<td align="left">Classes and Methods for Spatial Data</td>
</tr>
</tbody>
</table>
<div id="rgdal-rgeos-packages" class="section level3">
<h3><span class="header-section-number">1.1.1</span> RGDAL &amp; RGEOS</h3>
<p>Every time I upgrade in any significant way, two R libraries seem to raise their ugly heads and scream like a spoiled child—rgdal and rgeos. Why do these two have to be SOOOO much of a pain? Why can’t we have a auto build of a binary with all the options in it for OSX? Who knows? I always feel like I get the fuzzy end of the lollipop with these two. Here is my latest approach for getting them going.</p>
<p>First you have to make sure you have the latest GDAL libraries. I used to get mine from <a href="http://www.kyngchaos.com/software/frameworks">Kyngchaos</a>, just download the framework, install it, and then do some kind of long R CMD INSTALL dance, which seems to no longer work for me. I also tried installing from Ripley’s repository and found that (a) It was a version older than the one I already had on my machine, and (b) you can’t install from that repository, there is a malformed header and the install.packages() function just barfs.</p>
<p>Time to try something new. I typically stay away from the various installer frameworks out there on OSX to keep everything in Frameworks. But this time, I used MacPorts. You can find the latest version here. Here is how I got it to help me out.</p>
<ul>
<li>Download XCode from Apple, it is both free and has a lot of tools that make your life as a programmer easier. It is a big package and you’ll have to install the command line developer tools as well. You will be prompted on how to do this.</li>
<li>Downloaded the version of macports for your OS, I’m currently on 10.11 and installed it with little problems. It takes a bit of time at the end of the installation because it is downloading a lot of information. Be patient.</li>
<li>In the terminal, I updated it <code>sudo ports -v selfupdate</code> and again be patient, it is going to grab a lot of stuff from the internet.</li>
<li>I then used it to install <code>gdal</code> as a unix library (rather than as a framework so it won’t be located in /Library/Frameworks) by sudo ports install <code>gdal</code>. There were a lot of dependencies for this one so it took a while.</li>
<li>I then had R install rgdal as <code>install.packages( rgdal, type=&quot;source&quot;)</code></li>
</ul>
<p>Worked like a charm.</p>
</div>
</div>
<div id="data-sets" class="section level2">
<h2><span class="header-section-number">1.2</span> Data Sets</h2>
<p>This course will use two different data sets from the Dyer Laboratory.</p>
<div id="the-sonoran-desert-bark-beetle-araptus-attenuata" class="section level3">
<h3><span class="header-section-number">1.2.1</span> The Sonoran desert bark beetle, <em>Araptus attenuata</em></h3>
<p><img src="media/araptus.png" width="378" /></p>
<p><em>Arapatus attenuatus</em> is a bark beetle endemic to the Sonoran desert. It is known only from the senescing stems of the desert plant, <em>Euphorbia lomelii</em> (syn <em>Pedilanthus macrocarpus</em>). Samples for this data set were collected from 39 populations throughout the species range.</p>
<div class="figure">
<img src="LGDA_files/figure-html/unnamed-chunk-5-1.png" alt="(\#fig:unnamed-chunk-5)Sampling locales for the `arapat` data set." width="672" />
<p class="caption">
(#fig:unnamed-chunk-5)Sampling locales for the <code>arapat</code> data set.
</p>
</div>
<div id="nuclear-markers" class="section level4">
<h4><span class="header-section-number">1.2.1.1</span> Nuclear Markers</h4>
<p>The markers in the <code>arapat</code> data set consist of eight polymorphic and codominant nuclear markers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(gstudio)
<span class="kw">data</span>(arapat)
<span class="kw">column_class</span>( arapat, <span class="st">&quot;locus&quot;</span>) </code></pre></div>
<pre><code>## [1] &quot;LTRS&quot; &quot;WNT&quot;  &quot;EN&quot;   &quot;EF&quot;   &quot;ZMP&quot;  &quot;AML&quot;  &quot;ATPS&quot; &quot;MP20&quot;</code></pre>
<p>These markers span a range of allelic diversity and richness.</p>
<table>
<thead>
<tr class="header">
<th align="left">Locus</th>
<th align="center">A</th>
<th align="center">He</th>
<th align="center">Ae</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AML</td>
<td align="center">13</td>
<td align="center">0.829</td>
<td align="center">5.861</td>
</tr>
<tr class="even">
<td align="left">ATPS</td>
<td align="center">10</td>
<td align="center">0.719</td>
<td align="center">3.563</td>
</tr>
<tr class="odd">
<td align="left">EF</td>
<td align="center">2</td>
<td align="center">0.436</td>
<td align="center">1.774</td>
</tr>
<tr class="even">
<td align="left">EN</td>
<td align="center">5</td>
<td align="center">0.449</td>
<td align="center">1.815</td>
</tr>
<tr class="odd">
<td align="left">LTRS</td>
<td align="center">2</td>
<td align="center">0.499</td>
<td align="center">1.996</td>
</tr>
<tr class="even">
<td align="left">MP20</td>
<td align="center">19</td>
<td align="center">0.819</td>
<td align="center">5.512</td>
</tr>
<tr class="odd">
<td align="left">WNT</td>
<td align="center">5</td>
<td align="center">0.653</td>
<td align="center">2.880</td>
</tr>
<tr class="even">
<td align="left">ZMP</td>
<td align="center">2</td>
<td align="center">0.339</td>
<td align="center">1.514</td>
</tr>
</tbody>
</table>
</div>
<div id="methylation-markers" class="section level4">
<h4><span class="header-section-number">1.2.1.2</span> Methylation Markers</h4>
<p>In addition to codominant nuclear markers, an msAFLP analysis was performed on the major clade and paired sequence and <span class="math inline">\(C_pG\)</span> methylation markers were derived.</p>
</div>
</div>
<div id="the-flowering-dogwood-cornus-florida" class="section level3">
<h3><span class="header-section-number">1.2.2</span> The Flowering Dogwood <em>Cornus florida</em></h3>
<div class="figure">
<img src="media/dogwood_flower.png" alt="(\#fig:unnamed-chunk-8)Inflorescence for flowering dogwood with conspicuous showy bracts and many small flowers in the center." width="185" />
<p class="caption">
(#fig:unnamed-chunk-8)Inflorescence for flowering dogwood with conspicuous showy bracts and many small flowers in the center.
</p>
</div>
<p>Flowering dogwood is an ubiquitous feature of eastern north american forests.</p>
<p>The data set contains five microsatellite loci.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(cornus)
<span class="kw">column_class</span>( cornus, <span class="st">&quot;locus&quot;</span>) </code></pre></div>
<pre><code>## [1] &quot;Cf.G8&quot;  &quot;Cf.H18&quot; &quot;Cf.N5&quot;  &quot;Cf.N10&quot; &quot;Cf.O5&quot;</code></pre>
<p>of roughly equal genetic diversity.</p>
<table>
<thead>
<tr class="header">
<th align="left">Locus</th>
<th align="center">A</th>
<th align="center">He</th>
<th align="center">Ae</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Cf.G8</td>
<td align="center">28</td>
<td align="center">0.931</td>
<td align="center">14.482</td>
</tr>
<tr class="even">
<td align="left">Cf.H18</td>
<td align="center">20</td>
<td align="center">0.925</td>
<td align="center">13.325</td>
</tr>
<tr class="odd">
<td align="left">Cf.N10</td>
<td align="center">18</td>
<td align="center">0.903</td>
<td align="center">10.349</td>
</tr>
<tr class="even">
<td align="left">Cf.N5</td>
<td align="center">11</td>
<td align="center">0.453</td>
<td align="center">1.827</td>
</tr>
<tr class="odd">
<td align="left">Cf.O5</td>
<td align="center">15</td>
<td align="center">0.860</td>
<td align="center">7.146</td>
</tr>
</tbody>
</table>
<!--chapter:end:Configuring.Rmd-->
</div>
</div>
</div>
<div id="vector-raster-data" class="section level1 imageChapter">
<h1><span class="header-section-number">2</span> Vector &amp; Raster Data</h1>
<div class="chapter_image">
<img src="media/ch_newt2.jpg">
</div>
<p>Functionally, we will be working with two kinds of spatial data, vector and raster data. Vector data can be considered as a finite set of points that may or may not be connected. In <code>R</code>, these points can be used directly, as a numeric data type, or as a *Spatial** class object. The <code>sp</code> library contains a lot of functions that help deal with points, lines, and polygons and this is going to be a short overview of how they can be derived and manipulated in the pursuit of landscape genetic studies.</p>
<p>Rasters are a form of data that is georeferenced and (somewhat) more continuous. Raster data is perhaps best envisioned as a matrix of values, whose entries represent spatially referenced data points (pixels). The raster itself can be visualized as you could for normal matrix output. What makes a raster different, however, is that it is (or should be) georeferenced. This means that each element of the matrix represents some measurement on the ground having a specific location and spread—sometimes referred to as granularity. This is analogous to an image, where if you zoom in on it enough, you will be able to differentiate between individual pixels, each with its own color value. Raster representation specifies each pixel in terms of a location and size associated with that value that we can map onto the earth.</p>
<div id="points" class="section level2">
<h2><span class="header-section-number">2.1</span> Points</h2>
<p>Points are defined by <em>SpatialPoints</em> objects. A collection of points may have additional data associated with each location and would make a <em>SpatialPointsDataFrame</em>. This is a bit different than the normal <code>data.frame</code> objects we’ve been using with coordinates in them already—in fact it is the opposite of that. It is a set of points within which is located a <code>data.frame</code> rather than <code>data.frame</code> that has within it a set of points.</p>
<p>Confused yet? Lets get to the point and make some coordinates. Here is the way we’ve extracted points from the <em>Arapatus attenuatus</em> data set thus far.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(gstudio)
<span class="kw">data</span>(arapat)
coords &lt;-<span class="st"> </span><span class="kw">strata_coordinates</span>(arapat)
<span class="kw">summary</span>(coords)</code></pre></div>
<pre><code>##     Stratum     Longitude         Latitude    
##  101    : 1   Min.   :-114.3   Min.   :23.08  
##  102    : 1   1st Qu.:-112.9   1st Qu.:24.52  
##  12     : 1   Median :-111.5   Median :26.21  
##  153    : 1   Mean   :-111.7   Mean   :26.14  
##  156    : 1   3rd Qu.:-110.4   3rd Qu.:27.47  
##  157    : 1   Max.   :-109.1   Max.   :29.33  
##  (Other):33</code></pre>
<p>However, we can also derive these points directly as a <em>SpatialPoints</em> object defined in the <em>sp</em> library by setting the optional flag <code>as.SpatialPoints=TRUE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(sp)
<span class="kw">library</span>(raster)
pts &lt;-<span class="st"> </span><span class="kw">strata_coordinates</span>( arapat, <span class="dt">as.SpatialPoints =</span> <span class="ot">TRUE</span> )
pts</code></pre></div>
<pre><code>## class       : SpatialPoints 
## features    : 39 
## extent      : -114.2935, -109.1263, 23.0757, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : NA</code></pre>
<p>Notice that there is no coordinate reference system in the default extraction. This is because you can pass a wide array of coordinates to this function and it only takes the centroid. It is up to you to define the projection and datum for the data. If it is Long/Lat data as in the example, it can be defined as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">proj4string</span>(pts) &lt;-<span class="st"> </span><span class="kw">CRS</span>(<span class="st">&quot;+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0&quot;</span>)
pts</code></pre></div>
<pre><code>## class       : SpatialPoints 
## features    : 39 
## extent      : -114.2935, -109.1263, 23.0757, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0</code></pre>
<p>Any set of x- and y- coordinates can be turned into a <em>SpatialPoints</em> object. If we are to associate data with those points, the data has to have the same number of observations as there are coordinates. In our case here, we have 39 populations and as an example I’ll determine the number of individuals genotyped in each population as a</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">data.frame</span>( <span class="kw">table</span>(arapat$Population) )
<span class="kw">names</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Population&quot;</span>,<span class="st">&quot;N&quot;</span>)
pts.df &lt;-<span class="st"> </span><span class="kw">SpatialPointsDataFrame</span>(pts,df)
pts.df</code></pre></div>
<pre><code>## class       : SpatialPointsDataFrame 
## features    : 39 
## extent      : -114.2935, -109.1263, 23.0757, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0 
## variables   : 2
## names       : Population,  N 
## min values  :        101,  4 
## max values  :        SFr, 19</code></pre>
<p>You can translate it back into a <code>data.frame</code> object as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">as.data.frame</span>( pts.df )[<span class="dv">1</span>:<span class="dv">5</span>,]</code></pre></div>
<pre><code>##     Population  N         x        y
## 88         101  9 -114.2935 29.32541
## 9          102  8 -113.9449 29.01457
## 84          12 10 -113.6679 28.96651
## 175        153 10 -113.4897 28.72796
## 177        156  6 -113.9914 28.66056</code></pre>
<p>or access the data within the <code>data.frame</code> directly (thereby not needing to make a new object) using the attribute <code>@</code> operator</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pts.df@data[<span class="dv">1</span>:<span class="dv">5</span>,]</code></pre></div>
<pre><code>##   Population  N
## 1        101  9
## 2        102  8
## 3         12 10
## 4        153 10
## 5        156  6</code></pre>
<p>Since it is a <em>SpatialPoints</em> object, you can get information about it such as the bounding box (e.g., the coordinates of a box that encloses all the points).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bbox</span>(pts.df)</code></pre></div>
<pre><code>##         min        max
## x -114.2935 -109.12633
## y   23.0757   29.32541</code></pre>
</div>
<div id="lines" class="section level2">
<h2><span class="header-section-number">2.2</span> Lines</h2>
<p>Lines are created by pairs of points. A single <em>Line</em> object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">c1 &lt;-<span class="st"> </span><span class="kw">cbind</span>(coords$Longitude[<span class="dv">1</span>:<span class="dv">2</span>], coords$Latitude[<span class="dv">1</span>:<span class="dv">2</span>])
c2 &lt;-<span class="st"> </span><span class="kw">cbind</span>(coords$Longitude[<span class="dv">2</span>:<span class="dv">3</span>], coords$Latitude[<span class="dv">2</span>:<span class="dv">3</span>])
L1 &lt;-<span class="st"> </span><span class="kw">Line</span>(c1)
L2 &lt;-<span class="st"> </span><span class="kw">Line</span>(c2)
L1</code></pre></div>
<pre><code>## An object of class &quot;Line&quot;
## Slot &quot;coords&quot;:
##           [,1]     [,2]
## [1,] -114.2935 29.32541
## [2,] -113.9449 29.01457</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">coordinates</span>(L1)</code></pre></div>
<pre><code>##           [,1]     [,2]
## [1,] -114.2935 29.32541
## [2,] -113.9449 29.01457</code></pre>
<p>A collection of <em>Line</em> objects can be put into a <em>Lines</em> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ls1 &lt;-<span class="st"> </span><span class="kw">Lines</span>( <span class="kw">list</span>(L1), <span class="dt">ID=</span><span class="st">&quot;88 to 9&quot;</span>)
Ls2 &lt;-<span class="st"> </span><span class="kw">Lines</span>( <span class="kw">list</span>(L2), <span class="dt">ID=</span><span class="st">&quot;9 to 84&quot;</span>)
Ls1</code></pre></div>
<pre><code>## An object of class &quot;Lines&quot;
## Slot &quot;Lines&quot;:
## [[1]]
## An object of class &quot;Line&quot;
## Slot &quot;coords&quot;:
##           [,1]     [,2]
## [1,] -114.2935 29.32541
## [2,] -113.9449 29.01457
## 
## 
## 
## Slot &quot;ID&quot;:
## [1] &quot;88 to 9&quot;</code></pre>
<p>And if they are spatial in context (e.g., if we need to plot them in any way, shape, or form), we need to put them into a <em>SpatialLines</em> object, which is also constructed from a <code>list</code> of <em>Lines</em> objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SLs &lt;-<span class="st"> </span><span class="kw">SpatialLines</span>( <span class="kw">list</span>(Ls1,Ls2))
<span class="kw">proj4string</span>(SLs) &lt;-<span class="st"> </span><span class="kw">CRS</span>(<span class="kw">proj4string</span>(pts))
SLs</code></pre></div>
<pre><code>## class       : SpatialLines 
## features    : 2 
## extent      : -114.2935, -113.6679, 28.96651, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0</code></pre>
<p>If we want to add data to the set of lines, we can associate a <code>data.frame</code> with each of them with internal data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">data.frame</span>( <span class="dt">Sequence =</span> <span class="kw">c</span>(<span class="st">&quot;First&quot;</span>,<span class="st">&quot;Second&quot;</span>), <span class="dt">Like_It=</span> <span class="kw">c</span>(<span class="ot">TRUE</span>,<span class="ot">FALSE</span>), <span class="dt">row.names =</span> <span class="kw">c</span>(<span class="st">&quot;88 to 9&quot;</span>,<span class="st">&quot;9 to 84&quot;</span>))
SLDF &lt;-<span class="st"> </span><span class="kw">SpatialLinesDataFrame</span>( SLs, df )
SLDF</code></pre></div>
<pre><code>## class       : SpatialLinesDataFrame 
## features    : 2 
## extent      : -114.2935, -113.6679, 28.96651, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0 
## variables   : 2
## names       : Sequence, Like_It 
## min values  :    First,    TRUE 
## max values  :   Second,   FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">as.data.frame</span>(SLDF)</code></pre></div>
<pre><code>##         Sequence Like_It
## 88 to 9    First    TRUE
## 9 to 84   Second   FALSE</code></pre>
<p>We can also extract the line lengths of each line.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">SpatialLinesLengths</span>(SLs, <span class="dt">longlat =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## [1] 48.31561 27.51130</code></pre>
</div>
<div id="polygons" class="section level2">
<h2><span class="header-section-number">2.3</span> Polygons</h2>
<p>A polygon is simply a collection of line segments that closes in on itself. We can use polygons to identify habitat, define boundaries, etc. In the short description to follow, we will create a set Polygon* objects, culminating in a <em>SpatialPolygonsDataFrame</em> object.</p>
<p>We will start with the first 5 coordinates in the <code>arapat</code> data set. To make the polygon, we <strong>must</strong> close the coordinates, which means take the first one we put in and append it to the end of the list of coordinates, such that in this case <code>c[1,] == c[6,]</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">c &lt;-<span class="st"> </span><span class="kw">cbind</span>( coords$Longitude[<span class="dv">1</span>:<span class="dv">5</span>], coords$Latitude[<span class="dv">1</span>:<span class="dv">5</span>])
c &lt;-<span class="st"> </span><span class="kw">rbind</span>( c, c[<span class="dv">1</span>,])
P &lt;-<span class="st"> </span><span class="kw">Polygon</span>( c )
P</code></pre></div>
<pre><code>## An object of class &quot;Polygon&quot;
## Slot &quot;labpt&quot;:
## [1] -113.89179   28.89114
## 
## Slot &quot;area&quot;:
## [1] 0.1849395
## 
## Slot &quot;hole&quot;:
## [1] FALSE
## 
## Slot &quot;ringDir&quot;:
## [1] 1
## 
## Slot &quot;coords&quot;:
##           [,1]     [,2]
## [1,] -114.2935 29.32541
## [2,] -113.9449 29.01457
## [3,] -113.6679 28.96651
## [4,] -113.4897 28.72796
## [5,] -113.9914 28.66056
## [6,] -114.2935 29.32541</code></pre>
<p>As you can see, there is some additional information provided in the default layout. A few points to be made:<br />
- The <code>area</code> parameter is not georeferenced as the polygon itself has no projection.<br />
- The <code>labpt</code> is the coordinate where a label would be plot. - The <code>hole</code> and <code>ringDir</code> determine if the polygon represent a hole in some other polygon (e.g., the doughnut hole and the direction it is plot).</p>
<p>Similar to how we constructed <em>SpatialLines</em>, a <em>Polygon</em> must be in inserted into a set of <em>Polygons</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ps &lt;-<span class="st"> </span><span class="kw">Polygons</span>( <span class="kw">list</span>(P), <span class="dt">ID=</span><span class="st">&quot;Bob&quot;</span>)</code></pre></div>
<p>From which a list of can be created to make a <em>SpatialPolygons</em> object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SPs &lt;-<span class="st"> </span><span class="kw">SpatialPolygons</span>( <span class="kw">list</span>(Ps))
<span class="kw">proj4string</span>(SPs) &lt;-<span class="st"> </span><span class="kw">CRS</span>(<span class="kw">proj4string</span>(pts))
SPs</code></pre></div>
<pre><code>## class       : SpatialPolygons 
## features    : 1 
## extent      : -114.2935, -113.4897, 28.66056, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0</code></pre>
<p>And data can be added to it making a <em>SpatialPolygonsDataFrame</em> (n.b., The <code>row.names</code> of the <code>data.frame</code> <em>must</em> match the <code>ID</code> we set for making the <em>Polygons</em> objects). If they do not, there will be an error thrown.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Populations=</span><span class="kw">paste</span>(coords$Stratum[<span class="dv">1</span>:<span class="dv">5</span>],<span class="dt">collapse=</span><span class="st">&quot;, &quot;</span>), <span class="dt">row.names =</span> <span class="st">&quot;Bob&quot;</span>)
SPDF &lt;-<span class="st"> </span><span class="kw">SpatialPolygonsDataFrame</span>( SPs, df)
SPDF</code></pre></div>
<pre><code>## class       : SpatialPolygonsDataFrame 
## features    : 1 
## extent      : -114.2935, -113.4897, 28.66056, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0 
## variables   : 1
## names       :         Populations 
## min values  : 88, 9, 84, 175, 177 
## max values  : 88, 9, 84, 175, 177</code></pre>
</div>
<div id="rasters" class="section level2">
<h2><span class="header-section-number">2.4</span> Rasters</h2>
<p>Rasters are a form of data that is georeferenced and (somewhat) continuous. Raster data is perhaps best envisioned as a matrix of values, whose entries represent spatially referenced data points. The raster itself can be visualized as you could for normal matrix output. What makes a raster different, however, is that it is (or should be) georeferenced. This means that each element of the matrix represents some measurement on the ground having a specific location and spread. This is analogous to an image, where if you zoom in on it enough, you will be able to differentiate between individual pixels, it is just that for rasters, each pixel has a spatial location and size associated with it that we can map onto the earth.</p>
<p>You can either create raster objects <em>de novo</em> or you can acquire them from some external source. To create one from scratch, you start with a matrix of values and then construct the raster object using the <code>raster()</code> function as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(raster)
r &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">runif</span>(<span class="dv">10000</span>),<span class="dt">nrow=</span><span class="dv">100</span>)
rnd &lt;-<span class="st"> </span><span class="kw">raster</span>( r )</code></pre></div>
<p>which can be visualized using the normal plot command. The raster library has overridden several of the plotting functions and you can plot raster objects and decorate the images in the same way you do for normal plotting materials (@ref(graphics)).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(rnd)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-27-1.png" width="672" /></p>
<p>There are also many available repositories for raster data including Open Source, Governmental, and Municipal locations. One common source for these data is that of <a href="http://worldclim.org" class="uri">http://worldclim.org</a>. This repository contains temperature and precipitation data generalized for the entire globe.</p>
<p><img src="media/WorldClimTiles.png" width="333" /></p>
<p>These data are available free of charge and have been used in numerous biological studies. Moreover, they provide a set of ‘biologically relevant’ layers, called <code>BioClim</code>, that summarize both temperature and precipitation. They motivate these by saying:</p>
<blockquote>
<p>Bioclimatic variables are derived from the monthly temperature and rainfall values in order to generate more biologically meaningful variables. These are often used in ecological niche modeling (e.g., BIOCLIM, GARP). The bioclimatic variables represent annual trends (e.g., mean annual temperature, annual precipitation) seasonality (e.g., annual range in temperature and precipitation) and extreme or limiting environmental factors (e.g., temperature of the coldest and warmest month, and precipitation of the wet and dry quarters).</p>
</blockquote>
<p>These layers are encoded into 19 Bio-layers as defined in Table @ref(tab:bioclim). These layers are available for download from their site directly (I recommend using the tiles approach so you do not have to download the entire world map) as well as from the R package <a href="https://cran.r-project.org/web/packages/dismo/index.html">dismo</a>.</p>
<table>
<caption>
(#tab:bioclim)Key to the categories of bioclim variables derived from temperature and precipitation models for current (past and future) conditions. Rasters for these values (in GeoTiff and BIL formats) are available from <a href="http://worldclim.org" class="uri">http://worldclim.org</a>.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Layer
</th>
<th style="text-align:left;">
Description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
BIO1
</td>
<td style="text-align:left;">
Annual Mean Temperature
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO2
</td>
<td style="text-align:left;">
Mean Diurnal Range (Mean of monthly (max temp - min temp))
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO3
</td>
<td style="text-align:left;">
Isothermality (BIO2/BIO7 * 100)
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO4
</td>
<td style="text-align:left;">
Temperature Seasonality (standard deviation * 100)
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO5
</td>
<td style="text-align:left;">
Max Temperature of Warmest Month
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO6
</td>
<td style="text-align:left;">
Min Temperature of Coldest Month
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO7
</td>
<td style="text-align:left;">
Temperature Annual Range (BIO5-BIO6)
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO8
</td>
<td style="text-align:left;">
Mean Temperature of Wettest Quarter
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO9
</td>
<td style="text-align:left;">
Mean Temperature of Driest Quarter
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO10
</td>
<td style="text-align:left;">
Mean Temperature of Warmest Quarter
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO11
</td>
<td style="text-align:left;">
Mean Temperature of Coldest Quarter
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO12
</td>
<td style="text-align:left;">
Annual Precipitation
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO13
</td>
<td style="text-align:left;">
Precipitation of Wettest Month
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO14
</td>
<td style="text-align:left;">
Precipitation of Driest Month
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO15
</td>
<td style="text-align:left;">
Precipitation Seasonality (Coefficient of Variation)
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO16
</td>
<td style="text-align:left;">
Precipitation of Wettest Quarter
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO17
</td>
<td style="text-align:left;">
Precipitation of Driest Quarter
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO18
</td>
<td style="text-align:left;">
Precipitation of Warmest Quarter
</td>
</tr>
<tr>
<td style="text-align:left;">
BIO19
</td>
<td style="text-align:left;">
Precipitation of Coldest Quarter
</td>
</tr>
</tbody>
</table>
<p>For the purposes of this chapter, I’ll use bioclim and altitude layers from tile 22, which encompasses the spatial distribution of sampling locations in Baja California for the <em>Araptus attenuatus</em> dataset</p>
<p>Common raster formats include GeoTiff, essentially an image file with some metadata associated with it, and BIL (Binary interleaved) file formats. Both of these types are available from WorldClim. In general, the GeoTiff format is slightly easier to work with as all the data is contained within a single file, whereas the BIL format has two files for each raster (the second file is a header file that has the spatial meta data associated with it). If you do use the BIL format, the file path you pass to <code>raster()</code> would be of the BIL file, not the header one.</p>
<p>From Worldclim, I downloaded the elevation raster for Tile 22 and can load it into R using the <code>raster()</code> function as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alt &lt;-<span class="st"> </span><span class="kw">raster</span>(<span class="st">&quot;./spatial_data/alt.tif&quot;</span>)
alt</code></pre></div>
<pre><code>## class       : RasterLayer 
## dimensions  : 3600, 3600, 12960000  (nrow, ncol, ncell)
## resolution  : 0.008333333, 0.008333333  (x, y)
## extent      : -120, -90, 0, 30  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0 
## data source : /Users/rodney/Documents/Teaching/Landscape Genetics/Workshop - Scottland/Landscape Genetics Data Analysis/spatial_data/alt.tif 
## names       : alt 
## values      : -202, 5469  (min, max)</code></pre>
<p>The <code>alt</code> object is summarized here. A couple of things should be pointed out here: - In total, this is an object with 12,960,000 entries in it!<br />
- The resolution of each ‘pixel’ in this representation is 0.008, which is about 30-arc seconds or ~1km. That means that each loation in the study area is represented by the same exact value as the surrounding square kilometer. Obviously, if you are working on processes whose spatial scale is relevant less than 1000 m<sup><sup>2</sup></sup>, this kind of data is going to be of little value to you. - The values within the matrix range from -202 upwards to 5469. This is in meters.</p>
<p>In addition, the raster has a spatial extent and a projection associated with it. For more information on projections see @ref(map-projections).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bbox</span>( alt )</code></pre></div>
<pre><code>##     min max
## s1 -120 -90
## s2    0  30</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">proj4string</span>( alt )</code></pre></div>
<pre><code>## [1] &quot;+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0&quot;</code></pre>
<p>This elevation raster looks like:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>( alt, <span class="dt">xlab=</span><span class="st">&quot;Longitude&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Latitude&quot;</span> )</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-31-1.png" width="672" /></p>
</div>
<div id="cropping-rasters" class="section level2">
<h2><span class="header-section-number">2.5</span> Cropping Rasters</h2>
<p>Just because we have a large raster does not mean that it is in your best interest to use the entire object. Much of the spatial analyses routines used in population genetics require measurements of intervening distance, either Euclidean or ecological. Many of the routines for estimation of these distances require the estimation of pairwise distance between <em>all</em> pixels. For our purposes, the <code>arapat</code> dataset does not occur throughout most of this map, so it is in our best interests to use only the portion of the raster relevant to our data rather than the entire thing.</p>
<p>Here is one way of going this. I first define an extent, which consists of a vector representing the coordinates for <code>xmin</code>, <code>xmax</code>, <code>ymin</code>, and <code>ymax</code> (in decimal degrees longitude and latitude). You then <code>crop()</code> the raster to that extent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e &lt;-<span class="st"> </span><span class="kw">extent</span>( <span class="kw">c</span>(-<span class="dv">115</span>,-<span class="dv">109</span>,<span class="dv">22</span>,<span class="dv">30</span>) )
baja_california &lt;-<span class="st"> </span><span class="kw">crop</span>( alt, e )
<span class="kw">plot</span>(baja_california, <span class="dt">xlab=</span><span class="st">&quot;Longitude&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;Latitude&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-32-1.png" width="672" /></p>
<p>Lets make this base map a bit more pretty by taking the altitude and estimating the slope of each pixel and the direction it is facing (aspect). From this, we can ‘shade’ the hills in the map giving it more of a relief view we commonly see in maps.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">slope &lt;-<span class="st"> </span><span class="kw">terrain</span>( baja_california, <span class="dt">opt=</span><span class="st">&quot;slope&quot;</span> )
aspect &lt;-<span class="st"> </span><span class="kw">terrain</span>( baja_california, <span class="dt">opt=</span><span class="st">&quot;aspect&quot;</span>)
baja &lt;-<span class="st"> </span><span class="kw">hillShade</span>( slope, aspect, <span class="dv">40</span>, <span class="dv">270</span> )
<span class="kw">plot</span>(baja, <span class="dt">xlab=</span><span class="st">&quot;Longitude&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;Latitude&quot;</span>, <span class="dt">legend=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-33-1.png" width="672" /></p>
<p>Onto this map, we can plot our populations. For this, we convert the raw coordinates into a <code>SpatialPoints</code> object and then overlay onto the map. I use two <code>points()</code> commands to make the symbol for each population.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(arapat)
coords &lt;-<span class="st"> </span><span class="kw">strata_coordinates</span>(arapat)
pts &lt;-<span class="st"> </span><span class="kw">SpatialPoints</span>( coords[,<span class="dv">2</span>:<span class="dv">3</span>], <span class="dt">proj4string =</span> <span class="kw">CRS</span>(<span class="kw">proj4string</span>(baja)))
<span class="kw">plot</span>(baja, <span class="dt">xlab=</span><span class="st">&quot;Longitude&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;Latitude&quot;</span>, <span class="dt">legend=</span><span class="ot">FALSE</span>)
<span class="kw">points</span>( pts, <span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>, <span class="dt">pch=</span><span class="dv">3</span>)
<span class="kw">points</span>( pts, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-34-1.png" width="672" /></p>
<div id="cropping-rasters-via-polygons" class="section level3">
<h3><span class="header-section-number">2.5.1</span> Cropping Rasters Via Polygons</h3>
<p>It is also possible to crop a raster with a more fine grained approach using a polygon. Here is an example using five points picked around the region of Loreto, BCS (I just grabbed these by looking at Google Earth). You define a polygon by a series of points, the last of which has to be identical to the first one so that the polygon is a closed object and not just a series of points on a crooked line…</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pts &lt;-<span class="st"> </span><span class="kw">rbind</span>( <span class="kw">c</span>(-<span class="fl">111.5</span>,<span class="fl">27.0</span>),
              <span class="kw">c</span>(-<span class="fl">112.4</span>,<span class="fl">26.7</span>),
              <span class="kw">c</span>(-<span class="fl">111.7</span>,<span class="fl">25.7</span>),
              <span class="kw">c</span>(-<span class="fl">111.1</span>,<span class="fl">25.4</span>),
              <span class="kw">c</span>(-<span class="fl">110.8</span>,<span class="fl">26.0</span>),
              <span class="kw">c</span>(-<span class="fl">111.5</span>,<span class="fl">27.0</span>) )
pts</code></pre></div>
<pre><code>##        [,1] [,2]
## [1,] -111.5 27.0
## [2,] -112.4 26.7
## [3,] -111.7 25.7
## [4,] -111.1 25.4
## [5,] -110.8 26.0
## [6,] -111.5 27.0</code></pre>
<p>From these points, we construct a <code>SpatialPolygons</code> object (see @ref{polygons} for more info on this convoluted construction) and then can overlay onto the map to make sure it in the correct vicinity (here we are eyeballing it a bit). For more on why this next line of code looks so crazy, see @ref(polygons).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(baja, <span class="dt">legend=</span><span class="ot">FALSE</span>)
<span class="kw">plot</span>(polys, <span class="dt">add=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-36-1.png" width="672" /></p>
<p>To use the polygon to crop the raster, we have to both remove the part of the raster that is not contained within the polygon (<code>mask</code>) and then cut down the remaining raster to change the bounding box to that representing the portion of the data that remains (<code>trim</code>). If you do not trim the raster, it will have the same amount of data associated with it as the previous raster (e.g., the underlying data matrix will have 960 rows and 720 columns) but the part that is masked will be represented by <code>NA</code> values. For many rasters, the data is held in-memory (see the entry for ’data source<code>in the summary above) and as such removing as much of a raster that is</code>NA` improves your ability to manipulate it better.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">loredo &lt;-<span class="st"> </span><span class="kw">trim</span>( <span class="kw">mask</span>( baja, polys ) )
<span class="kw">plot</span>(loredo, <span class="dt">xlab=</span><span class="st">&quot;Longitude&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;Latitude&quot;</span>)</code></pre></div>
<div class="figure">
<img src="LGDA_files/figure-html/unnamed-chunk-37-1.png" alt="(\#fig:unnamed-chunk-37)Extraction of region within polygon from the full Baja California raster." width="672" />
<p class="caption">
(#fig:unnamed-chunk-37)Extraction of region within polygon from the full Baja California raster.
</p>
</div>
</div>
<div id="cropping-rasters-via-convex-hull" class="section level3">
<h3><span class="header-section-number">2.5.2</span> Cropping Rasters Via Convex Hull</h3>
<p>An analysis common to modern population genetics is that of finding ecological distances between objects on a landscape. The estimation of pairwise distance derived from spatial data is a computationally intensive thing, one that if you are not careful will bring your laptop to its knees! One way to mitigate this data problem is to use a minimal amount raster area so that the estimation of the underlying distance graph can be done on a smaller set of points.</p>
<p>Cropping by a polygon like demonstrated in the previous example is a ‘by hand’ approach to estimating a box that roughly encompasses your data points. A more efficient one is one where you simply provide your coordinates and we can estimate a polygon that surrounds those coordinates with the minimal amount of wasted space. This is called a <em>Convex Hull</em>, which is kind of like a polygon that is created as if there was a rubbrerband fit around all your points. It is a minimal area that includes all of your points.</p>
<p>For this example, I’m going to use the populations found along the peninsula and find the minimal area encompassing those points.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(arapat)
coords &lt;-<span class="st"> </span><span class="kw">strata_coordinates</span>(arapat)
baja_coords &lt;-<span class="st"> </span>coords[ !(coords$Stratum %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;101&quot;</span>,<span class="st">&quot;102&quot;</span>,<span class="st">&quot;32&quot;</span>)), ]
baja_pts &lt;-<span class="st"> </span><span class="kw">SpatialPoints</span>( baja_coords[,<span class="dv">2</span>:<span class="dv">3</span>])
<span class="kw">plot</span>(baja, <span class="dt">legend=</span><span class="ot">FALSE</span>)
<span class="kw">plot</span>(baja_pts,<span class="dt">add=</span>T,<span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>)
<span class="kw">plot</span>(baja_pts,<span class="dt">add=</span>T,<span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>,<span class="dt">pch=</span><span class="dv">16</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-38-1.png" width="672" /></p>
<p>The methods for finding the hull and adding a buffer around it are found in the <code>rgeos</code> package. These are pretty easy functions to use and are very helpful. If you are having trouble installing the <code>rgeos</code> package from source, see @ref(rgdal-rgeos-packages).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hull &lt;-<span class="st"> </span><span class="kw">gConvexHull</span>(baja_pts)
<span class="kw">plot</span>(baja, <span class="dt">legend=</span><span class="ot">FALSE</span>)
<span class="kw">plot</span>(baja_pts,<span class="dt">add=</span>T,<span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>)
<span class="kw">plot</span>(baja_pts,<span class="dt">add=</span>T,<span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>,<span class="dt">pch=</span><span class="dv">16</span>)
<span class="kw">plot</span>(hull,<span class="dt">add=</span>T,<span class="dt">border=</span><span class="st">&quot;red&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-39-1.png" width="672" /></p>
<p>The function <code>gConvexHull()</code> returns an object of type <em>SpatialPolygons</em>, just like we created before. However, we now have a polygon that has each of our most ‘outward’ populations on the very edge of the polygon. It may be beneficial for us to add a buffer around this polygon.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(baja, <span class="dt">legend=</span><span class="ot">FALSE</span>)
<span class="kw">plot</span>(baja_pts,<span class="dt">add=</span>T,<span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>)
<span class="kw">plot</span>(baja_pts,<span class="dt">add=</span>T,<span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>,<span class="dt">pch=</span><span class="dv">16</span>)
<span class="kw">plot</span>(hull_plus_buffer, <span class="dt">add=</span>T, <span class="dt">border=</span><span class="st">&quot;red&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-40-1.png" width="672" /></p>
<p>Now, we can mask and trim it to include only the area of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(pop_hull, <span class="dt">legend=</span><span class="ot">FALSE</span>, <span class="dt">xlab=</span><span class="st">&quot;Longitude&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Latitude&quot;</span>)
<span class="kw">plot</span>(baja_pts,<span class="dt">add=</span>T,<span class="dt">col=</span><span class="st">&quot;darkred&quot;</span>,<span class="dt">pch=</span><span class="dv">16</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-41-1.png" width="672" /></p>
<p>This would be a great raster to start looking at ecological separation in since we have removed the extraneous data that would unintentionally cause problems with the distance estimators.</p>
</div>
</div>
<div id="modifying-rasters" class="section level2">
<h2><span class="header-section-number">2.6</span> Modifying Rasters</h2>
<p>We can modify rasters just as easily as we can crop them. They are matrices, after all. The square bracket indexing you use for matrices are just as effective as before.</p>
<p>In the next example, I mask the landscape based upon elevation. I create a copy of the original raster and then make everything whose elevation is less than 500m as missing data. Plotting this over the top of the original raster shows only locations where elevation exceeds this cutoff.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bc[ bc &lt;<span class="st"> </span><span class="dv">500</span> ] &lt;-<span class="st"> </span><span class="ot">NA</span>
<span class="kw">plot</span>( baja_california, <span class="dt">legend=</span><span class="ot">FALSE</span>, <span class="dt">col=</span><span class="st">&quot;darkgrey&quot;</span> )
<span class="kw">plot</span>( bc, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">legend=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-42-1.png" width="672" /></p>
</div>
<div id="extracting-point-data-from-rasters" class="section level2">
<h2><span class="header-section-number">2.7</span> Extracting Point Data From Rasters</h2>
<p>So far, rastes have been confined to representing a single static object. However, it is not uncommon to need to query a raster and find out the values at particular points. These points may be pre-defined or they may be dynamic (e.g., you need to point at a location on the map and determine the value there).</p>
<p>For queries of the first kind, we can use the <code>extract()</code> function. For this I downloaded the average temperature and precipitation rasters from Worldclim.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">baja_temp &lt;-<span class="st"> </span><span class="kw">raster</span>(<span class="st">&quot;./spatial_data/bio1.tif&quot;</span>)
baja_prec &lt;-<span class="st"> </span><span class="kw">raster</span>(<span class="st">&quot;./spatial_data/bio12.tif&quot;</span>)</code></pre></div>
<p>And then extract the values from each of these layers into the <code>coords</code> data we already have set up.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">coords$elevation &lt;-<span class="st"> </span><span class="kw">extract</span>( baja_california, coords[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)])
coords$mean_temp &lt;-<span class="st"> </span><span class="kw">extract</span>( baja_temp, coords[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)])
coords$mean_precip &lt;-<span class="st"> </span><span class="kw">extract</span>( baja_prec, coords[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)])
coords[<span class="dv">1</span>:<span class="dv">10</span>,]</code></pre></div>
<pre><code>##    Stratum Longitude Latitude elevation mean_temp mean_precip
## 1       88 -114.2935 29.32541       681       178         143
## 11       9 -113.9449 29.01457       361       195         148
## 20      84 -113.6679 28.96651       368       197         124
## 29     175 -113.4897 28.72796       240       205         106
## 36     177 -113.9914 28.66056       177       203         120
## 46     173 -112.8698 28.40846        26       223         102
## 56     171 -113.1826 28.22308       522       195         145
## 66      89 -113.3999 28.03661       290       203         117
## 76     159 -113.3161 27.52944        87       211         123
## 85     SFr -112.9640 27.36320       305       205         107</code></pre>
<p>A note should be made on the temperature and precipitation values. Temperature is denoted in tenths of a degree Celcius. Though it does get quite hot at times, it does not average 188°C at population 88! Similarly, the units for precipitation are mm (or tenths of centimeters if you will…).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggrepel)
coords &lt;-<span class="st"> </span>coords[ <span class="kw">order</span>(coords$Latitude),]
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>( coords, <span class="kw">aes</span>(<span class="dt">x=</span>Latitude,<span class="dt">y=</span>elevation)) +<span class="st"> </span><span class="kw">geom_line</span>(<span class="dt">color=</span><span class="st">&quot;lightgrey&quot;</span>) 
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Elevation (m)&quot;</span>)
p +<span class="st"> </span><span class="kw">geom_text_repel</span>(<span class="kw">aes</span>(<span class="dt">label=</span>Stratum), <span class="dt">color=</span><span class="st">&quot;red&quot;</span>) </code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-45-1.png" width="672" /></p>
<p>The package <code>ggrepel</code> provides a pseudo-smart labelling geometry for ggplot allowing you to have labels that are shifted around the points so as to maximize visibility.</p>
<p>For inquires of the second type, we can use the function <code>click()</code> to retrieve one of several outputs. Here is the help file that describes the various components.</p>
<div class="scrollingbox">
<pre>
<pre>click {raster} R Documentation
Query by clicking on a map

Description

Click on a map (plot) to get values of a Raster* or Spatial* object at that location; and optionally the coordinates and cell number of the location. For SpatialLines and SpatialPoints you need to click twice (draw a box).

Usage

## S4 method for signature 'Raster'
click(x, n=Inf, id=FALSE, xy=FALSE, cell=FALSE, type="n", show=TRUE, ...)

## S4 method for signature 'SpatialGrid'
click(x, n=1, id=FALSE, xy=FALSE, cell=FALSE, type="n", ...)

## S4 method for signature 'SpatialPolygons'
click(x, n=1, id=FALSE, xy=FALSE, type="n", ...)

## S4 method for signature 'SpatialLines'
click(x, ...)

## S4 method for signature 'SpatialPoints'
click(x, ...)
Arguments

x   - Raster*, or Spatial* object (or missing)
n   - number of clicks on the map
id - Logical. If TRUE, a numeric ID is shown on the map that corresponds to the row number of the output
xy - Logical. If TRUE, xy coordinates are included in the output
cell - Logical. If TRUE, cell numbers are included in the output
type - One of "n", "p", "l" or "o". If "p" or "o" the points are plotted; if "l" or "o" they are joined by lines. See ?locator
show - logical. Print the values after each click?
... - additional graphics parameters used if type != "n" for plotting the locations. See ?locator

Value

The value(s) of x at the point(s) clicked on (or touched by the box drawn).

Note

The plot only provides the coordinates for a spatial query, the values are read from the Raster* or Spatial* object that is passed as an argument. Thus you can extract values from an object that has not been plotted, as long as it spatially overlaps with with the extent of the plot.

Unless the process is terminated prematurely values at at most n positions are determined. The identification process can be terminated by clicking the second mouse button and selecting 'Stop' from the menu, or from the 'Stop' menu on the graphics window.

See Also

select, drawExtent

Examples

r <- raster(system.file("external/test.grd", package="raster"))
#plot(r)
#click(r)
#now click on the plot (map)</pre>
</div>
<p>Here is the output from a single inquire on the <code>baja_california</code> raster map.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>( baja_california, <span class="dt">legend=</span><span class="ot">FALSE</span>)
<span class="kw">click</span>(baja_california, <span class="dt">xy=</span><span class="ot">TRUE</span>)
##           x        y value
## 1 -113.3875 27.82083   116</code></pre></div>
<div id="stacks-of-rasters" class="section level3">
<h3><span class="header-section-number">2.7.1</span> Stacks of Rasters</h3>
<p>It is not uncommon to be working with many different raster layers at the same time. Instead of loading them individually, the <code>raster</code> library has a <code>RasterStack</code> object that can hold several rasters at one time and can be used in places where we would use individual rasters. Here is an example using the elevation and temperature rasters for tile 22.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">files &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;./spatial_data/alt.tif&quot;</span>, <span class="st">&quot;./spatial_data/bio1.tif&quot;</span>, <span class="st">&quot;./spatial_data/bio5.tif&quot;</span>, <span class="st">&quot;./spatial_data/bio6.tif&quot;</span>)
bio_layers &lt;-<span class="st"> </span><span class="kw">stack</span>( files )
bio_layers</code></pre></div>
<pre><code>## class       : RasterStack 
## dimensions  : 3600, 3600, 12960000, 4  (nrow, ncol, ncell, nlayers)
## resolution  : 0.008333333, 0.008333333  (x, y)
## extent      : -120, -90, 0, 30  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0 
## names       :  alt, bio1, bio5, bio6 
## min values  : -202,  -20,   54,  -83 
## max values  : 5469,  295,  424,  242</code></pre>
<p>Performing operations on a stack is as easy as performing them on individual layers. Here, I trim them to the hull defined above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e &lt;-<span class="st"> </span><span class="kw">extent</span>( <span class="kw">c</span>(-<span class="dv">115</span>,-<span class="dv">109</span>,<span class="dv">22</span>,<span class="dv">30</span>) )
bio_layers &lt;-<span class="st"> </span><span class="kw">crop</span>( bio_layers, e )
<span class="kw">plot</span>(bio_layers)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-48-1.png" width="672" /></p>
<p>From which values may be extracted using normal methods as outlined above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">extract</span>( bio_layers, baja_pts)
df &lt;-<span class="st"> </span>df[ !<span class="kw">is.na</span>(df[,<span class="dv">1</span>]),]
<span class="kw">head</span>(df)</code></pre></div>
<pre><code>##      alt bio1 bio5 bio6
## [1,] 681  178  331   48
## [2,] 361  195  346   61
## [3,] 368  197  348   62
## [4,] 240  205  355   68
## [5,] 177  203  352   71
## [6,]  26  223  372   81</code></pre>
<p>And visualized normally.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>( GGally )
<span class="kw">ggpairs</span>(df)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-50-1.png" width="672" /></p>
</div>
</div>
<div id="rasters-ggplot" class="section level2">
<h2><span class="header-section-number">2.8</span> Rasters &amp; ggplot</h2>
<p>As is the case with a lot of data types in R, there is a way to use the <code>ggplot</code> library to visualize rasters. Essentially, what you need to do is to transform your raster objects into <code>data.frame</code> objects for <code>ggplot</code>’s <code>geom_tile()</code> function. Here is an example.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>( <span class="kw">rasterToPoints</span>( baja_california )) 
<span class="kw">names</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Longitude&quot;</span>,<span class="st">&quot;Latitude&quot;</span>,<span class="st">&quot;Elevation&quot;</span>)
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>( df ) +<span class="st"> </span><span class="kw">geom_tile</span>( <span class="kw">aes</span>(<span class="dt">x=</span>Longitude,<span class="dt">y=</span>Latitude,<span class="dt">fill=</span>Elevation)) 
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">scale_fill_gradientn</span>( <span class="dt">colors=</span><span class="kw">c</span>(<span class="st">&#39;#a6611a&#39;</span>,<span class="st">&#39;#dfc27d&#39;</span>,<span class="st">&#39;#f5f5f5&#39;</span>,<span class="st">&#39;#80cdc1&#39;</span>,<span class="st">&#39;#018571&#39;</span>))
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">coord_equal</span>() +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;Longitude&quot;</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Latitude&quot;</span>) 
p</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-51-1.png" width="672" /></p>
<p>As usual, we can add additional information to the plot and as we would for any other <code>ggplot</code> object. Here I’ll add the populations and indicate if they have samples in them that are of one species (Pure) or have a mix of the two (Mixed).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">num.clades &lt;-<span class="st"> </span><span class="kw">colSums</span>( <span class="kw">table</span>(arapat$Species, arapat$Population) &gt;<span class="st"> </span><span class="dv">0</span> )
Stratum=<span class="kw">names</span>(num.clades)
Species=<span class="st"> </span><span class="kw">factor</span>( <span class="kw">c</span>(<span class="st">&quot;Pure&quot;</span>,<span class="st">&quot;Mixed&quot;</span>)[num.clades] )
tmp.df &lt;-<span class="st"> </span><span class="kw">data.frame</span>( Stratum, Species )
sites &lt;-<span class="st"> </span><span class="kw">merge</span>( coords, tmp.df )
p +<span class="st"> </span><span class="kw">geom_point</span>( <span class="kw">aes</span>(<span class="dt">x=</span>Longitude,<span class="dt">y=</span>Latitude, <span class="dt">shape=</span>Species), <span class="dt">size=</span><span class="dv">3</span>, <span class="dt">data=</span>sites )</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-52-1.png" width="672" /></p>
</div>
<div id="d-visualization" class="section level2">
<h2><span class="header-section-number">2.9</span> 3D Visualization</h2>
<p>It is also possible to visualize rasters in 3-space. The library <code>rasterVis</code> provides an interface to the <code>rgl</code> library to plot a surface. Once installed, these are easy to use for viewing surfaces. Here is an example using the elevation data we have been playing with.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rasterVis)
<span class="kw">plot3D</span>( baja_california , <span class="dt">zfac=</span><span class="fl">0.1</span>)</code></pre></div>
<p><img src="media/Baja_Plot3D.png" width="200" /></p>
<p>The <code>zfac</code> option in the plot is the amount to scale the z-axis (elevation) in relation to the x-axis and y-axis dimensions. It is a bit exagerated at <code>zfac=0.1</code> but you get the idea.</p>
<!-- TODO: Add Categorical Rasters -->
<!-- ## Categorical Rasters  -->
<!-- **ADD HERE** -->
</div>
<div id="saving-exporting-vector-raster-objects" class="section level2">
<h2><span class="header-section-number">2.10</span> Saving &amp; Exporting Vector &amp; Raster Objects</h2>
<p>As all of these objects are R objects, they can be saved to disk using the <code>save()</code> function, which makes them a *.rda object. If you have objects that take a bit of time to create, it is in your best interests to save them after creation and on subsequent analyses, just use the saved versions.</p>
<p>There are many situations where you need to save a raster you’ve manipulated. As these raster objects are R objects, you can save them directly to the file system using the <code>write()</code> as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write</span>( baja_california, <span class="dt">filename=</span><span class="st">&quot;baja_california.rda&quot;</span>)</code></pre></div>
<p>This will save the raster object to file exactly like it is in your R session. To load it back in you just use <code>load()</code> and it is returned just like it was. The benefit to saving these as R objects is that you do not need to change it at all to pick up where you left off.</p>
<p>You may also need to export the vector or raster data into a non-R format for external analyses. A common format for vector data is the ubiquitous ESRI shapefile.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rgdal)
<span class="kw">writeOGR</span>( <span class="dt">obj=</span>SPDF, <span class="dt">dsn=</span><span class="st">&quot;spatial_data&quot;</span>, <span class="dt">layer=</span><span class="st">&quot;SpatialPolyExample&quot;</span>, <span class="dt">driver=</span><span class="st">&quot;ESRI Shapefile&quot;</span>)</code></pre></div>
<p>For rasters, you use the <code>writeRaster()</code> function. The file extension is used to determine the file format used and R saves it automatically.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">writeRaster</span>( baja_california, <span class="dt">filename=</span><span class="st">&quot;baja_california.tif&quot;</span>)</code></pre></div>
<!--chapter:end:Vector_Raster.Rmd-->
</div>
</div>
<div id="map-projections" class="section level1 imageChapter">
<h1><span class="header-section-number">3</span> Map Projections</h1>
<div class="chapter_image">
<img src="media/ch_butterfly2.jpg">
</div>
<p>A spatial projection is a mathematical representation of a coordinate space used to identify geospatial objects. Because the earth is both non-flat and non-spheroid, we must use mathematical approaches to describe the shape of the earth in a coordinate space. We do this using an ellipsoid—a simplified model of the shape of the earth. Common ellipsoids include:</p>
<ul>
<li>NAD27 (North American Datum of 1927) based upon land surveys<br />
</li>
<li>NAD83 based upon satellite data measuring the distance of the surface of the earth to the center of the plant. This is also internationally known as GRS80 (Geodetic Reference System 1980) internationally.<br />
</li>
<li>WGS84 (World Geodetic System 1984) is a refinement of GRS80 done by the US military that was used in the development of GPS systems (and subsequently for all of us).</li>
</ul>
<p>Onto this ellipsoid, we must define a set of reference locations (in 3-space) called datum that help describe the precise shape of the surface.</p>
<div id="projections" class="section level2">
<h2><span class="header-section-number">3.1</span> Projections</h2>
<p>A projection onto an ellipsoid is a way of converting the spherical coordinates, such as longitude and latitude, into 2-dimensional coordinates we can use. There are three main types of approaches that have been used to develop various projections. (see wikipedia for some example imagery of different projections).</p>
<p>These include:</p>
<ul>
<li>Azimuthal: An approach in which each region of the earth is projected onto a plane tangential to the surface, typically at the pole or equator. Cylindrical: This approach projects the surface of the earth onto a cylinder, which is ‘unrolled’ like a large map. This approach ‘stretches’ distances in a east-west fashion, which is why Greenland looks so large…</li>
<li>Conic: Another ‘unrolling’ approach, though this time instead of a cylinder, it is projected onto a cone.</li>
</ul>
<p>All projections produce bias in area, distance, or shape (some do so in more than one), so there is no ‘optimal’ projection. To give you an idea of the consequences of these projections, I’ll use the United States map as an example and we can visualize how it is projected onto a 2-dimensional space using different projections.</p>
<div id="equatorial-projections" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Equatorial Projections</h3>
<p>These are projections centered on the Prime Meridian (<code>Longitude=0</code>)</p>
<p><em>Mercator Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(maps)
<span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;mercator&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/projections1-1.png" width="672" /></p>
<p><em>MollWeide Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;mollweide&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-58-1.png" width="672" /></p>
<p><em>Gilbert Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;gilbert&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-59-1.png" width="672" /></p>
<p><em>Cylequalarea Projection</em></p>
<p>Some projections require additional parameters, this one is based upon equally spaced and straight meridians, equal area, and true centered on a particular Latitude. I used the centroid of the US.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;cylequalarea&quot;</span>,<span class="dt">par=</span><span class="fl">39.83</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-60-1.png" width="672" /></p>
</div>
<div id="azimuth-projections" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Azimuth Projections</h3>
<p>These projections are centered on the North Pole with parallels making concentric circles. Meridians are equally spaced radial lines.</p>
<p><em>Orthographic Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;orthographic&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-61-1.png" width="672" /></p>
<p><em>Stereographic Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;orthographic&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-62-1.png" width="672" /></p>
<p><em>Perspective Projection</em></p>
<p>Here the parameter is the distance (in earth radii) the observer is looking.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;perspective&quot;</span>, <span class="dt">param=</span><span class="dv">8</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-63-1.png" width="672" /></p>
<p><em>Gnomonic Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;gnomonic&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-64-1.png" width="672" /></p>
</div>
<div id="polar-conic-projections" class="section level3">
<h3><span class="header-section-number">3.1.3</span> Polar Conic Projections</h3>
<p>Here projections are symmetric around the Prime Meridian with parallel as segments of concentric circles with meridians being equally spaced.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;conic&quot;</span>,<span class="dt">par=</span><span class="fl">39.83</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-65-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;lambert&quot;</span>,<span class="dt">par=</span><span class="kw">c</span>(<span class="dv">30</span>,<span class="dv">40</span>))</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-66-1.png" width="672" /></p>
</div>
<div id="miscellaneous-projections" class="section level3">
<h3><span class="header-section-number">3.1.4</span> Miscellaneous Projections</h3>
<p><em>Square Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;square&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-67-1.png" width="672" /></p>
<p><em>Hexagon Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;hex&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-68-1.png" width="672" /></p>
<p><em>Bicentric Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;bicentric&quot;</span>, <span class="dt">par=</span>-<span class="dv">98</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-69-1.png" width="672" /></p>
<p><em>Guyou Projection</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">map</span>(<span class="st">&quot;state&quot;</span>,<span class="dt">proj=</span><span class="st">&quot;guyou&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-70-1.png" width="672" /></p>
<p>There are a lot of ways to project a 3-dimensional surface onto a 2-dimensional representation. Be aware of what you are using and how you are using it when plotting materials.</p>
</div>
<div id="reprojecting-rasters" class="section level3">
<h3><span class="header-section-number">3.1.5</span> Reprojecting Rasters</h3>
<p>When working with rasters, we can reproject these onto other projections rather easily. Here is an example from the worldclim elevation tile we used previously (see @ref(cropping-rasters)).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(raster)
alt &lt;-<span class="st"> </span><span class="kw">raster</span>(<span class="st">&quot;./spatial_data/alt.tif&quot;</span>)
e &lt;-<span class="st"> </span><span class="kw">extent</span>( <span class="kw">c</span>(-<span class="dv">115</span>,-<span class="dv">109</span>,<span class="dv">22</span>,<span class="dv">30</span>) )
baja_california &lt;-<span class="st"> </span><span class="kw">crop</span>( alt, e )
baja_california</code></pre></div>
<pre><code>## class       : RasterLayer 
## dimensions  : 960, 720, 691200  (nrow, ncol, ncell)
## resolution  : 0.008333333, 0.008333333  (x, y)
## extent      : -115, -109, 22, 30  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0 
## data source : in memory
## names       : alt 
## values      : -202, 2263  (min, max)</code></pre>
<p>We can now project it to another projection, lets say Lambert Conic Conformal.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rgdal)
<span class="kw">projection</span>(baja_california)</code></pre></div>
<pre><code>## [1] &quot;+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">baja_lcc &lt;-<span class="st"> </span><span class="kw">projectRaster</span>( baja_california, <span class="dt">crs=</span><span class="st">&quot;+proj=lcc +lat_1=48 +lat_2=33 +lon_0=-100 +ellps=WGS84&quot;</span>)
baja_lcc</code></pre></div>
<pre><code>## class       : RasterLayer 
## dimensions  : 1060, 878, 930680  (nrow, ncol, ncell)
## resolution  : 845, 935  (x, y)
## extent      : -1610637, -868727, 2790826, 3781926  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=lcc +lat_1=48 +lat_2=33 +lon_0=-100 +ellps=WGS84 
## data source : in memory
## names       : alt 
## values      : -202, 2205.154  (min, max)</code></pre>
<p>These two projections influence the region as shown below.</p>
<p><img src="LGDA_files/figure-html/unnamed-chunk-73-1.png" width="672" /></p>
</div>
<div id="projecting-in-ggplot" class="section level3">
<h3><span class="header-section-number">3.1.6</span> Projecting In GGPlot</h3>
<p>As usual, there is probably a way to plot these values in <code>ggplot</code> to make the output just a little bit more awesome. Projections of data in <code>ggplot</code> displays can be manipulated by appending a <code>coord_*</code> object to the plot. Here are two examples using a mercator and azimuth equal area projection of the state maps.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
states &lt;-<span class="st"> </span><span class="kw">map_data</span>(<span class="st">&quot;state&quot;</span>)
map &lt;-<span class="st"> </span><span class="kw">ggplot</span>( states, <span class="kw">aes</span>(<span class="dt">x=</span>long,<span class="dt">y=</span>lat,<span class="dt">group=</span>group))
map &lt;-<span class="st"> </span>map +<span class="st"> </span><span class="kw">geom_polygon</span>( <span class="dt">fill=</span><span class="st">&quot;white&quot;</span>,<span class="dt">color=</span><span class="st">&quot;black&quot;</span>)
map &lt;-<span class="st"> </span>map +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;Longitude&quot;</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Latitude&quot;</span>)
map +<span class="st"> </span><span class="kw">coord_map</span>(<span class="st">&quot;mercator&quot;</span>) </code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-74-1.png" width="672" /></p>
<p>Conversely, we can plot it using the equal area Azimuth projection</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">map +<span class="st"> </span><span class="kw">coord_map</span>(<span class="st">&quot;azequalarea&quot;</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-75-1.png" width="672" /></p>
<p>or fisheye</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">map +<span class="st"> </span><span class="kw">coord_map</span>(<span class="st">&quot;fisheye&quot;</span>,<span class="dt">par=</span><span class="dv">3</span>)</code></pre></div>
<p><img src="LGDA_files/figure-html/unnamed-chunk-76-1.png" width="672" /></p>
<p>or any other projection available listed in the <code>mapproject()</code> function.</p>
</div>
</div>
<div id="coordinate-systems" class="section level2">
<h2><span class="header-section-number">3.2</span> Coordinate Systems</h2>
<p>In R, we typically are dealing with a combination of data that we’ve collected and that we’ve attained from some other provider. In most GIS applications, the coordinate systems we encounter are either:</p>
<ul>
<li>UTM (Universal Transverse Mercator) measuring the distance from the prime meridian for the x-coordinate and the distance from the equator (often called northing in the northern hemisphere) for the y-coordinate. These distances are in meters and the globe is divided into 60 zones, each of which is 6 degrees in width. Geographic coordinate systems use longitude and latitude. For historical purposes these are unfortunately reported in degrees, minutes, seconds, a temporal abstraction that is both annoying and a waste of time (IMHO).<br />
</li>
<li>Decimal degrees, while less easy to remember, are easier to work with in R.</li>
<li>State Planar coordinate systems are coordinate systems that each US State has defined for their own purposes. They are based upon some arbitrarily defined points of reference and another pain to use (IMHO). Given the differential in state area, some states are also divided into different zones. Maps you get from municipal agencies may be in this coordinate system. If your study straddles different zones or even state lines, you have some work ahead of you…</li>
</ul>
<p>It is best to use a system that is designed for your kind of work. Do not, for example, use a state plane system outside of that state as you have bias associated with the distance away from the origin. That said, Longitude/Latitude (decimal degrees) and UTM systems are probably the easiest to work with in R.</p>
<p>In R, we use rgdal to project points. Here I load in the coordinates of the populations in the <em>Arapatus attenuatus</em> data set and make a <em>SpatialPoints</em> object out of it. Setting the <code>proj4string()</code> here <strong>does not</strong> project the data, I am just specifying that the data are already in the lat/long WGS84 format.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(sp)
<span class="kw">library</span>(gstudio)
<span class="kw">data</span>(arapat)
coords &lt;-<span class="st"> </span><span class="kw">strata_coordinates</span>( arapat )
pts &lt;-<span class="st"> </span><span class="kw">SpatialPoints</span>( coords[,<span class="dv">2</span>:<span class="dv">3</span>] )
<span class="kw">proj4string</span>(pts) &lt;-<span class="st"> </span><span class="kw">CRS</span>(<span class="st">&quot;+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0&quot;</span>)
pts</code></pre></div>
<pre><code>## class       : SpatialPoints 
## features    : 39 
## extent      : -114.2935, -109.1263, 23.0757, 29.32541  (xmin, xmax, ymin, ymax)
## coord. ref. : +proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0</code></pre>
<p>The <code>CRS()</code> function holds the definition of the projection and interfaces between the PROJ.4 and RGDAL libraries. To project a set of data points into a new coordinate systems, we use <code>spTransform()</code> and pass it the definition of the new system to use.</p>
<div id="changing-datum" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Changing Datum</h3>
<p>We will first begin by looking at differences in the actual datum used to record the loation of plots. Here I compare the decimal Longitude/Latitude we’ve used thus far with that from the Universal Transverse Mercator (<a href="https://en.wikipedia.org/wiki/Universal_Transverse_Mercator_coordinate_system">UTM</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pts.utm &lt;-<span class="st"> </span><span class="kw">spTransform</span>(pts, <span class="kw">CRS</span>(<span class="st">&quot;+proj=utm +zone=12 +datum=WGS84&quot;</span>))
<span class="kw">summary</span>( pts.utm )</code></pre></div>
<pre><code>## Object of class SpatialPoints
## Coordinates:
##               min       max
## Longitude  180128  686925.2
## Latitude  2552540 3248545.0
## Is projected: TRUE 
## proj4string :
## [+proj=utm +zone=12 +datum=WGS84 +ellps=WGS84 +towgs84=0,0,0]
## Number of points: 39</code></pre>
<p>You can see the tranformations in the coordinate system by comparing the plots below. The relative position of each point is the same.</p>
<div class="figure">
<img src="LGDA_files/figure-html/unnamed-chunk-79-1.png" alt="(\#fig:unnamed-chunk-79)Population locations in (A) longitude and latitude and (B) Universal Transverse Mercator coordinate systems." width="672" />
<p class="caption">
(#fig:unnamed-chunk-79)Population locations in (A) longitude and latitude and (B) Universal Transverse Mercator coordinate systems.
</p>
</div>
<!--chapter:end:Map_Projections.Rmd-->
</div>
</div>
</div>
<div id="part-genetic-data" class="section level1 unnumbered">
<h1>(PART) Genetic Data</h1>
<!--chapter:end:part_day_2.Rmd-->
</div>
<div id="genetic-markers-basic-summary-statistics" class="section level1">
<h1><span class="header-section-number">4</span> Genetic Markers &amp; Basic Summary Statistics</h1>
<div id="marker-types" class="section level2">
<h2><span class="header-section-number">4.1</span> Marker Types</h2>
</div>
<div id="sampling" class="section level2">
<h2><span class="header-section-number">4.2</span> Sampling</h2>
</div>
<div id="allele-frequencies" class="section level2">
<h2><span class="header-section-number">4.3</span> Allele Frequencies</h2>
</div>
<div id="hardy-weinberg-equilibrium" class="section level2">
<h2><span class="header-section-number">4.4</span> Hardy-Weinberg Equilibrium</h2>
<!--chapter:end:Markers_Summary_Stats.Rmd-->
</div>
</div>
<div id="genetic-distance-diversity" class="section level1">
<h1><span class="header-section-number">5</span> Genetic Distance &amp; Diversity</h1>
<!--chapter:end:Distance_Diversity.Rmd-->
</div>
<div id="genetic-structure" class="section level1">
<h1><span class="header-section-number">6</span> Genetic Structure</h1>
<!--chapter:end:Genetic_Structure.Rmd-->
</div>
<div id="part-barriers-resistance" class="section level1 unnumbered">
<h1>(PART) Barriers &amp; Resistance</h1>
<!--chapter:end:part_day_3.Rmd-->
</div>
<div id="remote-sensing-lidar" class="section level1">
<h1><span class="header-section-number">7</span> Remote Sensing &amp; Lidar</h1>
<!--chapter:end:RemoteSensing.Rmd-->
</div>
<div id="barrier-detection" class="section level1">
<h1><span class="header-section-number">8</span> Barrier Detection</h1>
<!--chapter:end:Barriers.Rmd-->
</div>
<div id="resistance-approaches" class="section level1">
<h1><span class="header-section-number">9</span> Resistance Approaches</h1>
<!--chapter:end:Resistance.Rmd-->
</div>
<div id="part-spatial-structure" class="section level1 unnumbered">
<h1>(PART) Spatial Structure</h1>
<!--chapter:end:part_day_4.Rmd-->
</div>
<div id="networks" class="section level1">
<h1><span class="header-section-number">10</span> Networks</h1>
<!--chapter:end:Networks.Rmd-->
</div>
<div id="autocorrelation" class="section level1">
<h1><span class="header-section-number">11</span> Autocorrelation</h1>
<!--chapter:end:Autocorrelation.Rmd-->
</div>
<div id="part-adaptive-variance" class="section level1 unnumbered">
<h1>(PART) Adaptive Variance</h1>
<!--chapter:end:part_day_5.Rmd-->
</div>
<div id="adaptive-variance" class="section level1">
<h1><span class="header-section-number">12</span> Adaptive Variance</h1>
<!--chapter:end:AdaptiveVariance.Rmd-->
</div>
<div id="quantitative-genetics" class="section level1">
<h1><span class="header-section-number">13</span> Quantitative Genetics</h1>
<!--chapter:end:QuantitativeGenetics.Rmd-->
</div>
<div id="genetic-distance-diversity-1" class="section level1">
<h1><span class="header-section-number">14</span> Genetic Distance &amp; Diversity</h1>
<!--chapter:end:ChromosomeWalking.Rmd-->
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->

<!--bookdown:config-->

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
